### PR TITLE
feat: add support for firestore queries

### PR DESCRIPTION
**Pull request by JVMartyns**
### **PR Type**
Enhancement, Tests

___

### **Description**
- Implement `run_query` function for Firestore queries.

- Add extensive test coverage for `run_query`.

- Introduce Elixir syntax for building Firestore queries.

- Update documentation and project configuration.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>api.ex</strong><dd><code>Implement `run_query` function for Firestore</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

lib/pplus_firestore/api.ex

<li>Added <code>run_query</code> function for Firestore queries.<br> <li> Utilized <code>RunQueryRequest</code> and <code>StructuredQuery</code>.

</details>

  </td>
  <td><a href="https://github.com/PagoPlus/pplus-firestore/pull/9/files#diff-cf8b2c1bf7d8eeff8cf0a31e84f4b7459c0cd8aedf15900ae41217dfaa607105">+16/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>decoder.ex</strong><dd><code>Support decoding of Firestore query responses</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

lib/pplus_firestore/decoder.ex

<li>Added decoding for <code>RunQueryResponse</code>.<br> <li> Enhanced decoding logic for query responses.

</details>

  </td>
  <td><a href="https://github.com/PagoPlus/pplus-firestore/pull/9/files#diff-1a0cac796cb8fa1e04140458a80a5cb3d91fdd15bf210bd51d1f578213644d01">+8/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>encoder.ex</strong><dd><code>Refactor and enhance Firestore data encoding</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

lib/pplus_firestore/encoder.ex

<li>Refactored <code>encode</code> and <code>encode_value</code> functions.<br> <li> Improved encoding logic for Firestore data.

</details>

  </td>
  <td><a href="https://github.com/PagoPlus/pplus-firestore/pull/9/files#diff-222f855b7be7ea4b2a2d2728ef1f5d599a551aaff638298524580561c51197c7">+16/-14</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>query.ex</strong><dd><code>Introduce Elixir syntax for Firestore queries</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

lib/pplus_firestore/query.ex

<li>Introduced Elixir syntax for Firestore queries.<br> <li> Supported various query operations and filters.

</details>

  </td>
  <td><a href="https://github.com/PagoPlus/pplus-firestore/pull/9/files#diff-177470f11f2342490772074f102c77fe2f4974ebe756001e5c7815d0fceb7653">+386/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>repo.ex</strong><dd><code>Add `run_query` function to repository</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

lib/pplus_firestore/repo.ex

<li>Added <code>run_query</code> function to the repository.<br> <li> Integrated query execution with Firestore.

</details>

  </td>
  <td><a href="https://github.com/PagoPlus/pplus-firestore/pull/9/files#diff-31892ed0ae79384a8861e28081474654ff64cd0889fc513c643a6e2a7d863470">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>mix.exs</strong><dd><code>Update documentation configuration in mix.exs</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

mix.exs

<li>Updated documentation configuration.<br> <li> Added image copying for documentation.

</details>

  </td>
  <td><a href="https://github.com/PagoPlus/pplus-firestore/pull/9/files#diff-dfa6f4ed74c90e5d4fda283d547d366586e690387289bcfd473e3fa5f9ace308">+8/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>api_test.exs</strong><dd><code>Add tests for Firestore `run_query` function</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

test/pplus_firestore/api_test.exs

<li>Added comprehensive tests for <code>run_query</code>.<br> <li> Covered various query scenarios and edge cases.

</details>

  </td>
  <td><a href="https://github.com/PagoPlus/pplus-firestore/pull/9/files#diff-ba89acd3b3b561ce3c9e17df4f87670646110103aaef7058758621eeda10e966">+503/-0</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about PR-Agent usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

### CODE REVIEW

File: lib/pplus_firestore/repo.ex
Comment (fermuch) [inline lib/pplus_firestore/repo.ex:1]: Adicionar `@spec` para ter disponível ao importar a macro

File: test/pplus_firestore/api_test.exs
Comment (fermuch) [inline test/pplus_firestore/api_test.exs:1]: Adicionar teste para verificar que é possível consultar uma sub-collection

File: lib/pplus_firestore/query.ex [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+  from("books", where: "author" == "John Doe", limit: 5, order_by: {"author", :asc})
+  ```
+
+  ```elixir
+  from("books", allDescendants: true)
+  ```
+  """
+  @spec from(collection :: String.t(), opts :: Keyword.t()) :: StructuredQuery.t()
+  defmacro from(collection, opts \\ []) do
+    all_descendants = Keyword.get(opts, :allDescendants, false)
```
Comment (fermuch) [inline lib/pplus_firestore/query.ex:89]: pensando aqui: não seria mais normal para alguém usando elixir usar `all_descendants` ou só `descendants`?

File: test/pplus_firestore/api_test.exs
Comment (fermuch) [inline test/pplus_firestore/api_test.exs:1]: não gosto da idea de estar repassando o `token` sempre. Seria melhor ter algo como um `%PPlusFireStore.Connection{}`, para manter o token lá e adicionar possíveis metadatos no futuro também.

**Review by fermuch — CHANGES_REQUESTED** [review]

File: test/pplus_firestore/api_test.exs
Comment (JVMartyns) [reply test/pplus_firestore/api_test.exs:1]: Sobre isso, tem o Goth que guarda o token no cache e busca um novo quando o último expira.
E tem o módulo PPlusFireStore.Repo que passa o token para a API.

**Review by fermuch — APPROVED** [review]
LGTM!
